### DISCUSSION TITLE
Handling missing value?

### DISCUSSION THREAD

**Discussion by mikko-ahonen** [EvgSkv/logica#363] [Q&A]
I am still struggling to implement predicate logic statement "every" with Logica.

For example:

"Every Employee who is a Manager has had ManagementTraining"

i.e.

Every Employee(x), Manager(x) -> ManagementTraining(x).

Because implication (Manager(x) -> ManagementTraining(x)) can be rewritten as ~Manager(x) or ManagementTraining(x), and I get the offending rows by negation of that, then I can check if there are zero rows the statement is true (1) or otherwise false (0).

So in SQL:

```
SELECT CASE WHEN c > 0 THEN 0 else 1 END AS r
FROM (
  SELECT COUNT(*) AS c
  FROM employee
  WHERE id IN (SELECT id from Manager) AND NOT id IN (SELECT id FROM ManagementTraining));
```

In Logica, I can do:

```
@Engine("sqlite");

Employee(id:"Bob");
Manager(id:"Bob");
ManagementTraining(id:"Bob");

Employee(id:"Charles");
Manager(id:"Charles");
ManagementTraining(id:"Charles");

Employee(id:"Alice");
Manager(id:"Alice");
ManagementTraining(id:"Alice");

Goal(count? += 1) distinct :-
  Employee(id:),
  Manager(id:),
  ~ManagementTraining(id:);
```

However, as count is implemented with sum, this gives me either the count of the offending rows but it gives None if there are no matches.

How do I turn this into 1 if c is None or 0 if c is defined in Logica?

Or is there an easier way to implement the statement in a way that is more close to the original predicate logic statement?

**Comment by EvgSkv** [comment]
First let's get discuss how we get rid of `null`. There are two recommended ways.

**Aggregating expressions**

```
Goal(count:) :-
  count_or_null += (1 :- Employee(id:), Manager(id:), ~ManagementTraining(id:)),
  count = Coalesce(count_or_null, 0);
```

This is probably most intuitive, but it is compiled to nested querries and SQL engine (especially SQLite) may get confused by it.

**Zero padding**

Zero padding pattern ends up useful in many situations in Logica. Here you make sure that aggregated column is never null by adding zero to all the keys that you expect to have. In this case there are no keys, there is a single row.

```
# Zero padding.
Goal(count? += 0) distinct;
# Actual value summation.
Goal(count? += 1) distinct :-
  Employee(id:),
  Manager(id:),
  ~ManagementTraining(id:);
```

Now the easest way to turn number greater than 0 to 1 is just by applying `Least` function, which returns least of values.

```
EndGoal(count: Least(count, 1)) :- Goal(count:);
```

Tested: [COLAB](https://colab.research.google.com/drive/17KQ4SDPs0ukjLPRy4ixupttcGXSdyydl?usp=sharing)

Let me know if you have any further questions, or desire for any improvement of this approach.

**Comment by mikko-ahonen** [comment]
I was able to solve the issues based on this, thank you very much!

I think it would be great if there would be generic built-ins for predicate logic "every x",  "there exists x" and implication.
